src/main/cpp/RLBotInterface/BallPrediction/src/primitives.cpp

```diff
@@ -0,0 +1,390 @@
+#include "..\inc\primitives.h"
+
+bool intersect(const interval & a, const interval & b) {
+  return (a.minimum <= b.maximum) && (a.maximum >= b.minimum);
+}
+
+bool intersect(const aabb & a, const aabb & b) {
+  return ((a.min_x <= b.max_x) & (a.max_x >= b.min_x) & 
+          (a.min_y <= b.max_y) & (a.max_y >= b.min_y) & 
+          (a.min_z <= b.max_z) & (a.max_z >= b.min_z));
+}
+
+template <>
+interval project_along_axis(const aabb & a, const vec3 & axis) {
+
+  vec3 abs_axis{fabs(axis[0]), fabs(axis[1]), fabs(axis(2))};
+  vec3 center = 0.5f * vec3{a.max_x+a.min_x, a.max_y+a.min_y, a.max_z+a.min_z};
+  vec3 diagonal{a.max_x - a.min_x, a.max_y - a.min_y, a.max_z - a.min_z};
+
+  float i_center = dot(axis, center);
+  float i_radius = 0.5f * dot(abs_axis, diagonal);
+
+  return interval{i_center - i_radius, i_center + i_radius};
+
+}
+
+template <>
+interval project_along_axis(const obb & a, const vec3 & axis) {
+
+  vec3 raxis = dot(axis, a.orientation); 
+  vec3 abs_raxis{fabs(raxis[0]), fabs(raxis[1]), fabs(raxis(2))};
+
+  float i_center = dot(axis, a.center); 
+  float i_radius = dot(abs_raxis, a.half_width);
+
+  return interval{i_center - i_radius, i_center + i_radius};
+
+}
+
+template <>
+interval project_along_axis(const tri & a, const vec3 & axis) {
+
+  float value;
+  interval i;
+
+  value = dot(a.p[0], axis);
+  i.minimum = i.maximum = value;
+
+  value = dot(a.p[1], axis);
+  i.minimum = fminf(value, i.minimum);
+  i.maximum = fmaxf(value, i.maximum);
+
+  value = dot(a.p[2], axis);
+  i.minimum = fminf(value, i.minimum);
+  i.maximum = fmaxf(value, i.maximum);
+
+  return i;
+
+}
+
+template < typename S, typename T >
+bool separated_along_axis(const S & a, const T & b, const vec3 & axis) {
+  auto a_interval = project_along_axis(a, axis);
+  auto b_interval = project_along_axis(b, axis);
+  return (intersect(a_interval, b_interval) == false);
+}
+
+bool intersect(const aabb & a, const obb & b) {
+  
+  vec3 e1{1.0f, 0.0f, 0.0f};
```

Comment:
can we expand these variable names?
If they are meaningless is it possible to give them meaning?